R/data_type.R

```diff
@@ -324,3 +341,82 @@ collapse_drugs <- function(df) {
   }
   df  
 }
+
+
+
+
+#' Cleanup additional perturbations in the data.table
+#' 
+#' This function processes drug and concentration columns in a data.table.
+#' It checks if there is only one unique drug (excluding a specified untreated tag)
+#' and if there are exactly two doses (one of which is 0). If these conditions are met,
+#' it creates a new column named after the drug and fills it with the doses,
+#' then removes the original drug and concentration columns.
+#'
+#' @param dt A data.table containing the data.
+#' @param drugs_cotrt_ids A vector of column names related to drugs.
+#' @param conc_cotrt_ids A vector of column names related to concentrations.
+#' @param untreated_tag A string representing the untreated tag (default is "vehicle").
+#' @return A modified data.table with new columns for the drugs and removed original drug and concentration columns.
+#' @examples
+#' library(data.table)
+#' dt <- data.table(
+#'   drug1 = c("vehicle", "drugA", "drugA"),
+#'   conc1 = c(0, 10, 0),
+#'   drug2 = c("vehicle", "drugB", "drugB"),
+#'   conc2 = c(0, 20, 0)
+#' )
+#' drugs_cotrt_ids <- c("drug1", "drug2")
+#' conc_cotrt_ids <- c("conc1", "conc2")
+#' dt <- process_drug_doses(dt, drugs_cotrt_ids, conc_cotrt_ids)
+#' print(dt)
+#' @export
+process_perturbations <- function(dt,
+                                  drugs_cotrt_ids,
+                                  conc_cotrt_ids,
+                                  untreated_tag = "vehicle") {
+  
+  # Assertions
+  checkmate::assert_data_table(dt)
+  checkmate::assert_character(drugs_cotrt_ids, any.missing = FALSE)
+  checkmate::assert_character(conc_cotrt_ids, any.missing = FALSE)
+  checkmate::assert_true(length(drugs_cotrt_ids) == length(conc_cotrt_ids))
+  
+  
+  # If lengths of drugs_cotrt_ids and conc_cotrt_ids are 0, return dt unchanged
+  if (length(drugs_cotrt_ids) == 0 && length(conc_cotrt_ids) == 0) {
+    return(dt)
+  } else {
+    
+    # Iterate through each pair of columns in drugs_cotrt_ids and conc_cotrt_ids
+    for (i in seq_along(drugs_cotrt_ids)) {
+      drug_col <- drugs_cotrt_ids[i]
+      conc_col <- conc_cotrt_ids[i]
+      
+      # Check if there is only one drug in the current drug column (excluding untreated_tag)
+      unique_drugs <- unique(dt[[drug_col]])
+      unique_drugs <- unique_drugs[!unique_drugs %in% untreated_tag]
+      
+      if (length(unique_drugs) == 1) {
+        # Check if there are only two doses in the current concentration column (0 and another value)
+        unique_doses <- unique(dt[[conc_col]])
+        
+        if (length(unique_doses) == 2 && 0 %in% unique_doses) {
+          # Create a new column named after the drug (excluding untreated_tag) and fill it with the doses
+          new_column_name <- unique_drugs[1]
+          dt[, (new_column_name) := dt[[conc_col]]]
+          
+          drug_order <- gsub(".*_(\\d+)$|.*", "\\1", drug_col)
```

Comment:
Add a test to ensure that this chunks works as expected.

Reply:
done